fix long error messages in nix-instantiate

narang99 opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

- Passing nix-expression using CLI puts the whole nix-expression in the error message making it unnecessarily verbose
- We pass it using stdio now instead of directly passing the expression as arg, this simply prints the command that failed


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

## Steps done in testing
- `nix-build -A tests.nixpkgs-check-by-name` for building it after making the required edits
- First ran on a normal checkout of nixpkgs, output: `Validation successful` (run command: ./result/bin/nixpkgs-check-by-name .`)
- removed `default.nix` to trigger error and print the stack tracing
- got a stack trace without the expression
```
hariomnarang@QURELAP-148: ~/Desktop/nixpkgs hariom/257748/check-by-name-msg!
$ ./result/bin/nixpkgs-check-by-name .                                              [18:08:26]
error:
       … while calling anonymous lambda

         at «stdin»:9:1:

            8| # - is_derivation: The result of `lib.isDerivation <attr>`
            9| {
             | ^
           10|   attrsPath,

       error: getting status of '/Users/hariomnarang/Desktop/nixpkgs/default.nix': No such file or directory
I/O error:  Failed to run command "nix-instantiate" "--eval" "--json" "--strict" "--readonly-mode" "--restrict-eval" "--show-trace" "--arg" "attrsPath" "/private/var/folders/s0/xjqwn7_d6kg3y6dcjy2tzpn80000gp/T/.tmpKXlpN1" "-I" "/private/var/folders/s0/xjqwn7_d6kg3y6dcjy2tzpn80000gp/T/.tmpKXlpN1" "--arg" "nixpkgsPath" "/Users/hariomnarang/Desktop/nixpkgs" "-I" "/Users/hariomnarang/Desktop/nixpkgs" "-"
```

Mic92 commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -63,21 +63,29 @@ pub fn check_values(
         .stderr(process::Stdio::inherit())
         // Clear NIX_PATH to be sure it doesn't influence the result
         .env_remove("NIX_PATH")
+        .stdin(process::Stdio::from(echo_out))
```
```suggestion
        .stdin(Stdio::piped())
```

Mic92 commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -96,6 +104,7 @@ pub fn check_values(
     }
+    command.arg("-");
 
     let result = command
         .output()
```
```suggestion
    let mut child = command.spawn();
    
    let child_stdin = child.stdin.as_mut().unwrap();
    child_stdin.write_all(EXPR)?;
    // Close stdin to finish and avoid indefinite blocking
    drop(child_stdin);
    
    let result = child.wait_with_output()
```

Mic92 commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -63,21 +63,29 @@ pub fn check_values(
     // With restrict-eval, only paths in NIX_PATH can be accessed, so we explicitly specify the
     // ones needed needed
 
+    let echo_expr_for_piping = process::Command::new("echo")
+        .arg(EXPR)
+        .stdout(process::Stdio::piped())
+        .spawn()
+        .expect("failed to echo command for piping to nix-instantiate");
+    let echo_out = echo_expr_for_piping
+        .stdout
+        .expect("failed to open echo stdout for piping to nix-instantiate");
+
```
```suggestion
```

Mic92 reviewed (dismissed):
Thanks for the fix. Instead of using the echo command, it feels more straight forward to just write to the pipe directly. While this can in theory block, the default pipe buffer should be big enough for our relatively short input do be submitted in just one write system call. If that wasn't the case we would have to worry about the stack size used to pass argv as well.

narang99 commented:
@Mic92 added the changes

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -96,10 +97,19 @@ pub fn check_values(
+        .spawn()
+        .unwrap_or_else(|err| panic!("failed to run command '{:?}': {}", stringify!(command), err));
+    let mut stdin = child.stdin.take().expect("Failed to open stdin");
+    std::thread::spawn(move || {
+        stdin
+            .write_all(EXPR.as_bytes())
+            .expect("Failed to write to stdin");
+    });
```
I originally suggested doing it this way, but this turns out to be a bit unfortunate, having to spawn a process just to write a constant file. I thought a bit about this and I think it could be nicer by letting Nix write the file to the `/nix/store` and passing its path to the program, which we can do by changing the `default.nix` like this:

```diff
diff --git a/pkgs/test/nixpkgs-check-by-name/default.nix b/pkgs/test/nixpkgs-check-by-name/default.nix
index 0d4e2fcc48a7..006f6971c337 100644
--- a/pkgs/test/nixpkgs-check-by-name/default.nix
+++ b/pkgs/test/nixpkgs-check-by-name/default.nix
@@ -4,9 +4,13 @@
   nix,
   rustfmt,
   clippy,
+  makeWrapper,
   mkShell,
 }:
 let
+
+  runtimeExprPath = "${./src/eval.nix}";
+
   package =
     rustPlatform.buildRustPackage {
       name = "nixpkgs-check-by-name";
@@ -16,7 +20,9 @@ let
         nix
         rustfmt
         clippy
+        makeWrapper
       ];
+      env.NIX_CHECK_BY_NAME_EXPR_PATH = runtimeExprPath;
       # Needed to make Nix evaluation work inside the nix build
       preCheck = ''
         export TEST_ROOT=$(pwd)/test-tmp
@@ -34,7 +40,12 @@ let
         cargo fmt --check
         cargo clippy -- -D warnings
       '';
+      postInstall = ''
+        wrapProgram $out/bin/nixpkgs-check-by-name \
+          --set NIX_CHECK_BY_NAME_EXPR_PATH "$NIX_CHECK_BY_NAME_EXPR_PATH"
+      '';
       passthru.shell = mkShell {
+        env.NIX_CHECK_BY_NAME_EXPR_PATH = runtimeExprPath;
         inputsFrom = [ package ];
       };
     };
```

This uses:
- `.env` to set `NIX_CHECK_BY_NAME_EXPR_PATH` for the build and the shell
- `wrapProgram --set` to propagate the environment variable to the resulting binary

Having this environment variable available in the program, you can pass its value as the argument instead of `-`.

Not only does this allow getting rid of the `spawn`, but it also makes it much easier to reproduce the command, since people can can the exact same expression via the store path in the argument :D

narang99 replied:
Yea sounds good, I'll make the changes. One thing, can you describe what `makeWrapper` does? I tried looking through it but am not very used to jumping around nix code right now (not very good with nix also :( ).   
Just need to know the intention of that specific part, so that I can figure out what the code is doing easily. The whole `wrapProgram` thing is slightly confusing, on first sight, I probably would have simply set the env and hoped the executable would be called with it...

infinisil replied:
Ah yes, there are [docs for `wrapProgram`](https://nixos.org/manual/nixpkgs/unstable/#fun-makeWrapper), which just calls `makeWrapper` (the shell function, not the Nix expression) underneath, for which there are also [docs](https://nixos.org/manual/nixpkgs/unstable/#fun-makeWrapper). Unfortunately it's not very well documented. The source is better in this case. While the docs mention the source file, I can also recommend using a command like

```
❯ nix-shell -p ripgrep

[nix-shell:~/src/nixpkgs]$ rg 'makeWrapper\(\)'
```

to locate shell commands in Nixpkgs. In this case it points to https://github.com/NixOS/nixpkgs/blob/5d6c73222280b7f69109dfa5c9572f4b74c3453c/pkgs/build-support/setup-hooks/make-wrapper.sh#L10-L35

In short: `wrapProgram foo <args>` replaces the binary `foo` with a shell wrapper that calls the original `foo` underneath, but with some changes according to `<args>`. In this case, I used `--env` to make it have an environment variable set when it runs.

We need this because the build runs in a completely isolated environment, which is shut down after the build is done. After the build, the binary is by default run with whatever environment variables the callers process has, which could be anything, even completely empty!

So if we want the program to also have some environment variable when it runs, it needs to be explicitly propagated from the build.

---

Admittedly this is still not ideal, because it should be possible to embed the Nix store path of the expression into the Rust binary itself, such that it doesn't need to read via an environment variable. This could be done with something like [`include_str!("expr-path")`](https://doc.rust-lang.org/std/macro.include_str.html), where you'd do something like this in the `default.nix` file:

```nix
preConfigure = ''
  echo -n "${runtimeExprPath}" > src/expr-path
'';
```

But this is also not that great because it makes it harder for the development shell to work correctly.

narang99 replied:
Done, thanks for the help. One last thing, can you help me understand `passthru.shell`? I saw the definition for `passthru` in nix manual, the definition itself is trivial and does not really give an intuition on what is going around with that.   
I tested it by removing the environment variable from the expression and it does not come up in `nix-shell`. How the nix-shell is interacting with passthru is not very clear to me from the definition [here](https://nixos.org/manual/nixos/stable/#test-opt-passthru) alone.

narang99 replied:
@infinisil I've added the fix, can you review it please?

infinisil reviewed (changes requested):
Awesome, thanks a lot! I've left some suggestions :)

hariom-narang-qure committed:
pass eval.nix as a file instead of expression

- passing it as expression gives large error messages
  which are not very readable
- this commits puts the file in nix-store
  and patches the final program to have access to
  the path to the file as env.
- We simply pass this file to nix-instantiate

infinisil reviewed (changes requested):
Looking pretty good!

hariom-qure commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -60,9 +58,9 @@ pub fn check_values(
         attrs_file_path.display()
     ))?;
 
+    let expr_path = std::env::var("NIX_CHECK_BY_NAME_EXPR_PATH")?;
```
I was confused about this, thought we wanted the path to be linked at build time (compile time), although runtimepath does suggest otherwise lol, do we want runtime path to be linked so that development is easy? Isn't it better to simply nix run or something that takes care of it?

narang99 replied:
Was confused about this, I thought we wanted to link it during compile time and keep everything in the store? We want to link it and still keep it outside store?

Im assuming it's to make development easy, but isn't it cleaner to do nix test or something for development?

infinisil replied:
I assume you meant to reply to https://github.com/NixOS/nixpkgs/pull/267048#discussion_r1397974041

We only need it in the store so that the result can always find it without problems. It can't be a local (non-store) file there, since there's no guarantees about what the local filesystem looks like when the result is run.

But during development, we _do_ know the local filesystem, there's always going to be a `./src/eval.nix` file relative to the project root, so we can just use that instead. Does that make sense?

narang99 replied:
Yea. Works.

Commit:
made EXPR_PATH point to local instead of store

- we need this behavior in dev shell
- the main build still uses the store file

narang99 commented:
@infinisil committed your suggestions

infinisil reviewed (approved):
Looking good to me, thanks a lot!
